src/lib/tools.ts

```diff
@@ -0,0 +1,245 @@
+		uri,
+	);
+}
+
+export function localPathURI(uri: string): boolean {
+	return /^([a-zA-Z0-9_/\-.])+$/.test(uri);
+}
+
+export async function loadFromURI(
+	URI: string,
+	path: string,
+	errorMessage?: string,
+): Promise<any> {
+	// TODO: support git URI
+	if (gitURI(URI)) {
+		throw new UnsupportedError('Git URI support not implemented yet');
+	}
+	try {
+		if (localPathURI(URI)) {
+			return await loadFromFile(join(URI, path), errorMessage);
+		}
+	} catch (e) {
+		throw new URILoadError(`Error loading ${path} from ${URI}\n${e.message}`);
+	}
+}
+
+export function getBasePath(path: string): string {
+	return dirname(resolve(path));
+}
+
+export function convertRelativePaths({
+	conf,
+	basePath,
+}: {
+	conf: any;
+	basePath: string;
+}): any {
+	// Convert relative to absolute URIs
+	const keys = [
+		'productRepo',
+		'archiveStore',
+		'encryptionKeyPath',
+		'envFile.path',
+		'yamlFile.path',
+		'kubernetes.kubeConfigPath',
+		'kubernetes.bastion.key',
+		'compose.socket',
+	];
+
+	for (const k of keys) {
+		const value = _.get(conf, k);
+		if (value) {
+			_.set(conf, k, getAbsoluteURI(value, basePath));
+		}
+	}
+	return conf;
+}
+
+export async function readFromURI(
+	URI: string,
+	path: string,
+	cachePath?: string,
+): Promise<any> {
+	// TODO: support git URI
+	if (gitURI(URI)) {
+		throw new UnsupportedError('Git URI support not implemented yet');
+	} else if (localPathURI(URI)) {
+		return (await fs.readFile(join(URI, path))).toString('utf8');
+	}
+}
+
+export async function listURI(
+	URI: string,
+	path: string,
+	cachePath?: string,
+): Promise<any> {
+	// TODO: support git URI
+	if (gitURI(URI)) {
+		throw new UnsupportedError('Git URI support not implemented yet');
+	} else if (localPathURI(URI)) {
+		return await fs.readdir(join(URI, path));
+	}
+}
+
+/**
+ * Keyframe unwrapper to a standard format
+ * @param productRepoURI: URI of the product repo
+ * @param keyFramePath: path of keyframe
+ *
+ * returns: Keyframe object
+ */
+export async function unwrapKeyframe(
+	productRepoURI: string,
+	keyFramePath: string = './keyframe.yml',
+): Promise<object> {
+	// TODO: keyframe layering
+	let keyFrame = await loadFromURI(productRepoURI, keyFramePath);
+
+	if (keyFrame) {
+		keyFrame = _.filter(
+			_.get(keyFrame, ['children', 'sw', 'containerized-application'], []),
+			component => {
+				return component.type === 'sw.containerized-application';
+			},
+		);
+		keyFrame = _.mapValues(_.keyBy(keyFrame, 'slug'), o => {
+			return _.merge(_.get(o, 'assets', {}), { version: _.get(o, 'version') });
+		});
+		return keyFrame;
+	} else {
+		return {};
+	}
+}
+
+/**
+ * Creates an ssh tunnel for executing a promise
+ * @param tnlConfig: ssh2 tunnel configuration object
+ * @param prom: promise
+ * @param timeout: tunnel timeout.
+ * @returns {Promise<any>}
+ */
+export async function runInTunnel(
+	tnlConfig: any,
+	prom: any,
+	timeout: number,
+): Promise<any> {
+	return await tunnelAsync(tnlConfig).then(tnl => {
```

Comment:
You can use `await` here for all of this without having to resort to a promise chain.